src/plugins/intel_cpu/src/nodes/subgraph.cpp

```diff
@@ -240,264 +212,345 @@ InferenceEngine::Precision Snippet::getRuntimePrecision() const {
+                           dataSize[i]);
     }
 }
-
-static auto collapseLastDims(std::vector<size_t>& dims, size_t dimsToCollapse) -> void {
-    if (dimsToCollapse >= dims.size() - 1)
-        IE_THROW() << "Got invalid number of dims to collapse. Expected < " << dims.size() - 1 << " got " << dimsToCollapse;
-    for (int i = dims.size() - 2; i > dims.size() - dimsToCollapse - 2; i--) {
-        dims[dims.size() - 1] *= dims[i];
-    }
-
-    for (int i = dims.size() - 2; i >= dimsToCollapse; i--) {
-        dims[i] = dims[i - dimsToCollapse];
-    }
-
-    for (int i = dimsToCollapse - 1; i >= 0; i--) {
-        dims[i] = 1;
+void Snippet::optimizeExecDomain(std::vector<VectorDims>& inputShapes, std::vector<VectorDims>& outputShapes,
+                                 VectorDims &domain, size_t& TileRank) const {
+    const size_t ds = domain.size();
+    if (ds <= 2) { // not enough dimensions to collapse
+        return;
     }
-}
-
-void Snippet::define_schedule() {
-    auto edgeToBlockedShape = [](const EdgePtr& edge) {
-        const auto blockedDesc = edge->getMemory().GetDescWithType<BlockedMemoryDesc>();
-        ngraph::Shape shape(blockedDesc->getBlockDims());
-        ngraph::AxisVector blocking(blockedDesc->getOrder());
-        ngraph::element::Type precision = InferenceEngine::details::convertPrecision(blockedDesc->getPrecision());
-        return ngraph::snippets::op::Subgraph::BlockedShape{shape, blocking, precision};
-    };
-    auto prependWithOnes = [this](const std::vector<size_t>& dims) {
-        if (tensorRank <= dims.size())
-            return dims;
-        VectorDims result(tensorRank, 1);
-        std::copy(dims.begin(), dims.end(), &result[tensorRank - dims.size()]);
-        return result;
-    };
-    ngraph::snippets::op::Subgraph::BlockedShapeVector input_blocked_shapes;
-    for (size_t i = 0; i < inputShapes.size(); i++)
-        input_blocked_shapes.push_back(edgeToBlockedShape(getParentEdgesAtPort(i)[0]));
-
-    ngraph::snippets::op::Subgraph::BlockedShapeVector output_blocked_shapes;
-    for (size_t i = 0; i < outputShapes.size(); i++)
-        output_blocked_shapes.push_back(edgeToBlockedShape(getChildEdgesAtPort(i)[0]));
 
-    exec_domain = snippet->canonicalize(output_blocked_shapes, input_blocked_shapes);
-
-    // initialize by maximum output dimension. Dimensions of outputs should be broadcastable
-    tensorRank = std::max(static_cast<size_t>(rank6D), exec_domain.size());
-    // Canonicalization broadcasts inputs and outputs to max input rank, which can be smaller than tensorRank
-    // prepend to enable 6D scheduler
-    exec_domain = prependWithOnes(exec_domain);
-    const auto &body = snippet->get_body();
-    for (const auto& p : body->get_parameters()) {
-        dims_in.emplace_back(prependWithOnes(p->get_shape()));
+    // There are operations which don't support dim collapsion but we can execute inside Tile 2D
+    const auto body = snippet->get_body()->get_ops();
```

Comment:
We can add a special key to snippets config `has_domain_sensitive_operations` and omit calling optimizeExecDomain() if it's true. Please see the transpose PR for details.